### PR TITLE
Apply version stamps with Nerdbank.GitVersioning

### PR DESCRIPTION
**Pull request by AArnott**
This replaces MinVer with [Nerdbank.GitVersioning](https://github.com/dotnet/Nerdbank.GitVersioning/).

It also applies the policy that `TypeShape.Roslyn` should be precisely versioned, which is one of the two requirements for safe consumption from roslyn analyzers and source generators.

BTW, an optional and useful tool to use in combination with this is [`nbgv`](https://github.com/dotnet/Nerdbank.GitVersioning/blob/main/doc/nbgv-cli.md), which can automate releases, tagging, etc.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -15,6 +15,8 @@ jobs:
 
     steps:
     - uses: actions/checkout@v4
+      with:
+        fetch-depth: 0 # avoid shallow clone so nbgv can do its work.
```
Comment (eiriktsarpalis) [inline .github/workflows/build.yml:19]: Does it need to clone the entire history to work? Can it be truncated to a reasonable upper bound perhaps?

File: Directory.Packages.props [hunk 2]
```diff
@@ -28,4 +27,7 @@
     <PackageVersion Include="JsonSchema.Net" Version="7.2.3" />
     <PackageVersion Include="FSharp.Core" Version="8.0.401" />
   </ItemGroup>
-</Project>
\ No newline at end of file
+  <ItemGroup>
+    <GlobalPackageReference Include="Nerdbank.GitVersioning" Version="3.6.146" />
```
Comment (eiriktsarpalis) [inline Directory.Packages.props:31]: I would probably only scope this to projects under the src/ folder

File: .github/workflows/build.yml (hunk 1 above)
Comment (AArnott) [reply .github/workflows/build.yml:19]: It only needs to be high enough to reach the last change to version.json. But in practice, that's a hard number to estimate, and unless you're in a repo like VS, shallow clones don't really save significant time on clone in my experience.

File: Directory.Packages.props (hunk 2 above)
Comment (AArnott) [reply Directory.Packages.props:31]: That's an option. But why bother to add nuance?

File: src/TypeShape.Roslyn/version.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+  "$schema": "https://raw.githubusercontent.com/dotnet/Nerdbank.GitVersioning/main/src/NerdBank.GitVersioning/version.schema.json",
+  "inherit": true,
+  "assemblyVersion": {
```
Comment (eiriktsarpalis) [inline src/TypeShape.Roslyn/version.json:4]: Presumably this is refining the assembly version for `TypeShape.Roslyn` specifically? What is the default assembly version for the others?

Comment (AArnott) [reply src/TypeShape.Roslyn/version.json:4]: The default is to take major.minor from the `version` property for the assembly version.

**Pull request merged**